# NLP - Emotion Classification in Text
## Objective: 
Develop machine learning models to classify emotions in text samples.

### 1. Loading and Preprocessing (3 marks)
Load the dataset and perform necessary preprocessing steps. This should include text cleaning, tokenization, and removal of stopwords. Explain the preprocessing techniques used and their impact on model performance.

In [1]:
import pandas as pd
import re
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
import nltk

# Download NLTK resources
nltk.download('punkt')
nltk.download('stopwords')

# Load the dataset
data = pd.read_csv('nlp_dataset.csv')

# Display the first few rows of the dataset
print(data.head())


[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\gokul\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\gokul\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


                                             Comment Emotion
0  i seriously hate one subject to death but now ...    fear
1                 im so full of life i feel appalled   anger
2  i sit here to write i start to dig out my feel...    fear
3  ive been really angry with r and i feel like a...     joy
4  i feel suspicious if there is no one outside l...    fear


In [2]:
# Define a function for text cleaning
def clean_text(text):
    # Remove punctuation and numbers
    text = re.sub(r'[^a-zA-Z\s]', '', text)
    # Convert to lowercase
    text = text.lower()
    return text

# Apply text cleaning
data['cleaned_comment'] = data['Comment'].apply(clean_text)


In [3]:
# Tokenization and stopword removal
stop_words = set(stopwords.words('english'))

def preprocess_text(text):
    tokens = word_tokenize(text)
    tokens = [word for word in tokens if word not in stop_words]
    return ' '.join(tokens)

data['processed_comment'] = data['cleaned_comment'].apply(preprocess_text)

# Display the processed comments
print(data[['Comment', 'processed_comment', 'Emotion']].head())

                                             Comment  \
0  i seriously hate one subject to death but now ...   
1                 im so full of life i feel appalled   
2  i sit here to write i start to dig out my feel...   
3  ive been really angry with r and i feel like a...   
4  i feel suspicious if there is no one outside l...   

                                   processed_comment Emotion  
0  seriously hate one subject death feel reluctan...    fear  
1                         im full life feel appalled   anger  
2  sit write start dig feelings think afraid acce...    fear  
3  ive really angry r feel like idiot trusting fi...     joy  
4  feel suspicious one outside like rapture happe...    fear  


### 2. Feature Extraction (2 marks):
Implement feature extraction using CountVectorizer or TfidfVectorizer. Describe how the chosen method transforms the text data into numerical features.

In [4]:
from sklearn.feature_extraction.text import TfidfVectorizer
# Initialize TfidfVectorizer
vectorizer = TfidfVectorizer()

# Fit and transform the processed comments into TF-IDF features
X = vectorizer.fit_transform(data['processed_comment'])
y = data['Emotion']

### 3. Model Development (2 marks):
Train the following machine learning models

## Naive Bayes Model

In [5]:
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score, f1_score

# Split the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Initialize and train the Naive Bayes model
nb_model = MultinomialNB()
nb_model.fit(X_train, y_train)

# Make predictions and evaluate the model
nb_predictions = nb_model.predict(X_test)
nb_accuracy = accuracy_score(y_test, nb_predictions)
nb_f1_score = f1_score(y_test, nb_predictions, average='weighted')

print(f'Naive Bayes Accuracy: {nb_accuracy:.2f}, F1 Score: {nb_f1_score:.2f}')

Naive Bayes Accuracy: 0.91, F1 Score: 0.91


## Support Vector Machine Model

In [6]:
from sklearn.svm import SVC

# Initialize and train the SVM model
svm_model = SVC(kernel='linear')
svm_model.fit(X_train, y_train)

# Make predictions and evaluate the model
svm_predictions = svm_model.predict(X_test)
svm_accuracy = accuracy_score(y_test, svm_predictions)
svm_f1_score = f1_score(y_test, svm_predictions, average='weighted')

print(f'SVM Accuracy: {svm_accuracy:.2f}, F1 Score: {svm_f1_score:.2f}')

SVM Accuracy: 0.95, F1 Score: 0.95


### 4. Model Comparison (2 marks)
Evaluate the model using appropriate metrics (e.g., accuracy, F1-score). Provide a brief explanation of the chosen model and its suitability for emotion classification.

In [7]:
# Print comparison results
print("\nModel Comparison:")
print(f"Naive Bayes - Accuracy: {nb_accuracy:.2f}, F1 Score: {nb_f1_score:.2f}")
print(f"SVM - Accuracy: {svm_accuracy:.2f}, F1 Score: {svm_f1_score:.2f}")


Model Comparison:
Naive Bayes - Accuracy: 0.91, F1 Score: 0.91
SVM - Accuracy: 0.95, F1 Score: 0.95


## Model Comparison

After training accuracy and F1 score of models 
- *Naive Bayes*
  - *Accuracy**: 0.91
  - *F1 Score*: 0.91
- *Support Vector Machine (SVM)*
  - *Accuracy*: 0.95
  - *F1 Score*: 0.95
## Conclusion

 the SVM model have higher accuracy and F1 score than  the Naive Bayes model , suggesting it is better at understanding and classifying emotions in text data. This means that for tasks like identifying feelings expressed in writing, SVM may be a more reliable choice than Naive Bayes based on these results.